src/sdk/object.hpp

```diff
@@ -0,0 +1,616 @@
+/**
+ MARK: Embedded Object
+ `EmbeddedObject` is a base class used to define embedded Realm model objects.
+
+ Embedded objects work similarly to normal objects, but are owned by a single
+ parent Object (which itself may be embedded). Unlike normal top-level objects,
+ embedded objects cannot be directly created in or added to a Realm. Instead,
+ they can only be created as part of a parent object, or by assigning an
+ unmanaged object to a parent object's property. Embedded objects are
+ automatically deleted when the parent object is deleted or when the parent is
+ modified to no longer point at the embedded object, either by reassigning an
+ Object property or by removing the embedded object from the List containing it.
+
+ Embedded objects can only ever have a single parent object which links to
+ them, and attempting to link to an existing managed embedded object will throw
+ an exception.
+
+ The property types supported on `EmbeddedObject` are the same as for `Object`,
+ except for that embedded objects cannot link to top-level objects, so `Object`
+ and `std::vector<Object*>` properties are not supported (`EmbeddedObject` and
+ `std::vector<EmbeddedObject>` *are*).
+
+ Embedded objects cannot have primary keys or indexed properties.
+
+ ```cpp
+ struct Dog: EmbeddedObject<Dog> {
+    REALM(std::string) name;
+    REALM(bool) adopted;
+
+    REALM_EXPORT(name, adopted);
+ }
+ struct Owner: Object<Owner> {
+    REALM(std::string) name;
+    REALM(std::vector<Dog>) dogs;
+    REALM_EXPORT(name, dogs);
+ }
+ ```
+ */
+template <typename T>
+struct Property<T, util::type_traits::embedded_object>
+: public ObjectBase<T>
+{
+    using base = ObjectBase<T>;
+    using self = Object<T>;
+    using derived_t = T;
+    using base::base;
+    using AccessorBase::realm;
+
+    void attach_and_write(SharedRealm r, TableRef table, ObjKey key) override
+    {
+        base::attach(r, table, key);
+        if (AccessorBase::column_name.empty()) {
+            // if there is no column name, this is either in a list
+            // or a top level object
+            base::attach_and_write(realm(), table, key);
+        } else {
+            auto obj = AccessorBase::get_obj().create_and_set_linked_object(AccessorBase::get_column_key());
+            base::attach_and_write(realm(), obj.get_table(), obj.get_key());
+        }
+    }
+
+    operator T() const
+    {
+        return static_cast<const T&>(*this);
+    }
+
+    friend struct PropertyFriend;
+    friend base;
+};
+
+/// @see `Property<T, util::type_traits::embedded_object>`
+template <typename T>
+using EmbeddedObject = Property<T, util::type_traits::embedded_object>;
+
+// MARK: EmbeddedObjectAccessor
+template <typename T>
+struct Accessor<T, typename std::enable_if_t<util::is_embedded_v<T>>> : public AccessorBase
+{
+    using type_info = TypeInfo<T>;
+    using base = AccessorBase;
+    using base::base;
+    using decayed_t = std::decay_t<T>;
+
+    Accessor(const AccessorBase& o) : base(o)
```

Comment:
Do you intend this class to be implicitly convertible from any other `Accessor<T>`?